aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+
+  thrust::transform(
+      policy,
+      thrust::make_counting_iterator(int64_t(0)),
+      thrust::make_counting_iterator(int64_t(nnz)),
+      thrust::device_ptr<int64_t>(offsets.data_ptr<int64_t>()),
+      [indices_accessor, strides_ptr, dim, ndim] __device__(int64_t x) {
+        int64_t pool_index = 0;
+        for (int64_t j = 0; j < ndim; j++) {
+          if (j != dim) {
+            auto indices_row = indices_accessor[j];
+            auto stride = strides_ptr[j];
+            pool_index += stride * indices_row[x];
+          }
+        }
+        return pool_index;
+      });
+  return offsets;
+}
+
+template <class scalar_t, bool requireMxRows = true>
+std::tuple<Tensor, Tensor, Tensor, Tensor> compute_pool_max(
+    const Tensor& indices,
+    const Tensor& values,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
+  /*
+    Return pools of indices that align with the given dimension and the
+    correspoinding max values for each pool.
+
+    See ATen/native/sparse/cpu/Softmax.cpp:get_offsets and
+    ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation that this implementation is based on.
+  */
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto nnz = indices.size(1);
+  auto offsets  = get_offsets(indices, sizes, dim);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+
+  auto sorted_indices = at::empty({nnz}, indices.options());
+  thrust_ptr sorted_indices_thrust_ptr(sorted_indices.data_ptr<int64_t>());
+  thrust::sequence(
+      policy, sorted_indices_thrust_ptr, sorted_indices_thrust_ptr + nnz, 0);
+
+  thrust::sort(
+      policy,
+      sorted_indices_thrust_ptr,
+      sorted_indices_thrust_ptr + nnz,
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] < offsets_ptr[y];
+      });
+  auto pool_sizes = at::empty({nnz}, indices.options());
+
+  auto new_end = thrust::reduce_by_key(
+      policy,
+      sorted_indices_thrust_ptr,
+      sorted_indices_thrust_ptr + nnz,
+      thrust::make_constant_iterator(int64_t(1)),
+      thrust::make_discard_iterator(),
+      thrust_ptr(pool_sizes.data_ptr<int64_t>()),
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] == offsets_ptr[y];
+      });
+  auto new_sz = thrust::distance(
+      thrust_ptr(pool_sizes.data_ptr<int64_t>()), new_end.second);
+  pool_sizes.resize_({new_sz});
+
+  auto pool_offsets = pool_sizes.clone();
+  thrust_ptr pool_offsets_thrust_ptr(
+      pool_offsets.data_ptr<int64_t>());
+  thrust::exclusive_scan(
+      policy,
+      pool_offsets_thrust_ptr,
+      pool_offsets_thrust_ptr + new_sz,
+      pool_offsets_thrust_ptr);
+
+  Tensor mx_buffer;
+  if (requireMxRows) {
+    mx_buffer = at::empty({new_sz * nvalues}, values.options());
```

Comment:
Why not just create the tensor filled with the correct value (of negative infinity)?

Reply:
All right, so for the best of my knowledge this has to be done in two steps 
```
    auto output = at::empty(new_shape, self.options());
    output.fill_(value);
```
I will change using this approach. If there is a better way please let me know, cc @mruberry

Reply:
Can you use `full`? https://pytorch.org/docs/master/generated/torch.full.html?highlight=full#torch.full

Reply:
Yes I can